### ISSUE TITLE
osc hangs if origin and target are the same rank

### ISSUE THREAD

**Issue by rkowalewski** [open-mpi/ompi#2538]
I have provided a small [gist](https://gist.github.com/rkowalewski/c737ce4135fa3b1b6febd4767f1b220f#file-ompi-bug-reflexive-comm-c-L66) where you can verify this. The program cannot make any progress if origin and target are the same rank. Open-MPI v 1.10.x and other MPI libraries terminate successfully.

**Comment by rkowalewski** [comment]
Some additional information about my environment.

~~~bash
$ ompi_info
 Open MPI: 2.0.1
 Open MPI repo revision: v2.0.0-257-gee86e07
 Open MPI release date: Sep 02, 2016
 Open RTE: 2.0.1

$ mpicc --version
 gcc (GCC) 5.4.0
 Copyright (C) 2015 Free Software Foundation, Inc.
~~~

**Comment by rkowalewski** [comment]
One additional question: Is there any possibility to enable progress threads? Or are there configure flags to enable this feature at compilation?

**Comment by jsquyres** [comment]
@rkowalewski Just out of curiosity, does the same problem happen on the latest 2.0.x nightly snapshot tarball?  https://www.open-mpi.org/nightly/v2.0.x/

**Comment by fuchsto** [comment]
@jsquyres It's a bit tricky to deploy the nightly in our CI containers, but I am setting up a separate environment for OpenMPI-2 tests and will post my results.

**Comment by fuchsto** [comment]
@jsquyres
The nightly snapshot is looks promising indeed! Our internal CI passed flawlessly on the first try.
OpenMPI build has been configured with:

~~~
./configure --prefix=$HOME/opt/openmpi-2.0.1x \
            --enable-debug \
            --enable-mem-debug \
            --enable-mem-profile \
            --enable-pretty-print-stacktrace \
            --with-hwloc=internal
~~~

Apart fromt the fixes, performance also seems to have improved. Even with debugging options and no tuning attempts, overall execution time of our test suite is down by a stable 5% compared to the current release.
I will integrate the nightly snapshot in our CI and report the results here.

**Comment by rkowalewski** [comment]
@jsquyres I can confirm that this bug does not exist in the current nightly build anymore (Nightly build v2.0.x from 12/11/2016). The attached [gist](https://gist.github.com/rkowalewski/c737ce4135fa3b1b6febd4767f1b220f#file-ompi-bug-reflexive-comm-c-L66) termintes without any errors.

**Comment by jsquyres** [comment]
@rkowalewski Sweet!  Since it will be fixed in our upcoming 2.0.2 release, I'll go ahead and close this issue.  Thanks for following up!
